pkg/discovery/worker/garbage_collector.go

```diff
@@ -110,3 +160,144 @@ func (g *GarbageCollector) isLeakedPod(pod api.Pod) bool {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (g *GarbageCollector) unpauseRollouts() error {
+	for _, grandParentRes := range supportedGrandParents {
+		objList, err := g.dynClient.Resource(grandParentRes).Namespace("").List(context.TODO(), gcListOpts)
+		if err != nil {
+			// This error could be because of mising deploymentConfig type from the cluster, so we don't fail here
+			if apierrors.IsNotFound(err) && strings.Contains(err.Error(), "the server could not find the requested resource") {
+				glog.V(3).Infof("Resource not found enabled on cluster while resuming rollout of controllers %s: %v", grandParentRes.String(), err)
+				continue
+			}
+			glog.Errorf("Error resuming rollout of controller %s: %v", grandParentRes.String(), err)
+			return err
+		}
+		if objList == nil {
+			// ignore
+			return nil
+		}
+		for _, item := range objList.Items {
+			unpause := false
+			err := commonutil.RetryDuring(executor.DefaultRetryLess, 0,
+				executor.DefaultRetrySleepInterval, func() error {
+					return executor.ResourceRollout(g.dynClient.Resource(grandParentRes).Namespace(item.GetNamespace()), &item, unpause)
+				})
+			if err != nil {
+				glog.Errorf("Error resuming rollout of controller %s %s/%s: %v", item.GetKind(), item.GetNamespace(), item.GetName(), err)
+				return err
+			}
+		}
+
+	}
+
+	return nil
+}
+
+func (g *GarbageCollector) cleanupQuotas() error {
+	quotaList, err := g.client.CoreV1().ResourceQuotas("").List(context.TODO(), gcListOpts)
+	if err != nil {
+		glog.Errorf("Error garbage cleaning quotas: %v", err)
+		return err
+	}
+	if quotaList == nil {
+		// ignore
+		return nil
+	}
+
+	for _, quota := range quotaList.Items {
+		err := revertQuota(g.client, &quota)
+		if err != nil {
+			// we force kubeturbo to restart
+			return err
+		}
+	}
+
+	return nil
+}
+
+func revertQuota(client *kubernetes.Clientset, quota *api.ResourceQuota) error {
+	var revertedQuota *api.ResourceQuota
+	var err error
+	if quota.Annotations != nil {
+		origSpec, exists := quota.Annotations[executor.QuotaAnnotationKey]
+		if exists {
+			revertedQuota, err = executor.DecodeQuota([]byte(origSpec))
+			if err != nil {
+				// This is very unlikely but not an error which we can recover from, for example in the next run.
+				// We need to ignore this. The annotation if there still is any would then carry the opportunity
+				// for an user to use this as a reference in case a manual correction is required ever.
+				glog.Warningf("Error reverting quota while garbage collecting resources: %s/%s: %v", quota.Namespace, quota.Name, err)
+			}
+		}
+	}
+	if revertedQuota != nil {
+		// Although unlikely, revertedQuota being nil possibly the result of the decode error above
+		// so we go ahead and try to remove the GC label alone and leave the annotation behind.
+		quota.Spec.Hard = revertedQuota.Spec.Hard
+		RemoveTurboAnnotionFromQuota(quota)
+	}
+	executor.RemoveGCLabelFromQuota(quota)
```

Comment:
Hi @irfanurrehman, I may be missing here. I understand it would be a rare case to fail the decoding. But I'm wondering if we want to remove BOTH turbo annotation and GC label from quota only if decoding succeeds (revertedQuota != nil).

I noticed in your changes above in `Revert()` func in `quota_util`, when decoding fails, we'll continue in the for loop without calling `RemoveGCLabelFromQuota`:
```
func (q *QuotaAccessorImpl) Revert() {
	for _, quota := range q.quotas {
		var revertedQuota *corev1.ResourceQuota
		var err error
		if quota.Annotations != nil {
			origSpec, exists := quota.Annotations[QuotaAnnotationKey]
			if exists {
				revertedQuota, err = decodeQuota([]byte(origSpec))
				revertedQuota, err = DecodeQuota([]byte(origSpec))
				if err != nil {
					glog.Warningf("Error reverting the updated quota: %s/%s: %v", quota.Namespace, quota.Name, err)
					continue
				}
			}
		}
		if revertedQuota == nil {
			continue
		}

		RemoveGCLabelFromQuota(revertedQuota)
		...
	}
}
```

Or maybe to make it simple, we could just return the err when decoding fails above? Something like:
```
revertedQuota, err = executor.DecodeQuota([]byte(origSpec))
			if err != nil {
				// This is very unlikely but not an error which we can recover from, for example in the next run.
				// We need to ignore this. The annotation if there still is any would then carry the opportunity
				// for an user to use this as a reference in case a manual correction is required ever.
				glog.Warningf("Error reverting quota while garbage collecting resources: %s/%s: %v", quota.Namespace, quota.Name, err)
                                return err
                         }
```

Reply:
There are 2 cases I could think of the decode would fail.
1. Dependency encode/decode functionality is updated (in k8s) and there is some mismatch in the encode/decode that we are using such that the data encoded by the encode method cannot be decoded by our decode method.
2. After we place the annotation, another entity (for example an user) updates it before the decode is hit in the flow, such that the update breaks the expected json structure and is no longer valid for decode.

In both the above cases (or for that matter any other possible decode failure case), the failure won't be transient. That is it will always fail and there would be no recovery. 
Now I chose not to fail in the normal flow (`quota_util.go revert()`), because failing at that time would mean that we will skip updating rest of the quotas (if multiple quota resources were updated to allow the pod move) for a failure which we cannot anyways recover from. The label is left behind for it to be tried once more in the garbage collection flow (when ever kubturbo restarts again). In the garbage collection flow, we fail kubeturbo start on failure, so if we return failure on a non recoverable failure, kubeturbo will remain in an infinite restart loop and hence the comment in the snippet you posted:
```
// This is very unlikely but not an error which we can recover from, for example in the next run.
// We need to ignore this. The annotation if there still is any would then carry the opportunity
// for an user to use this as a reference in case a manual correction is required ever.
```

Does this make sense? Plz let me know if you still have concerns?